File: 576 Out of Boundary Paths/576 Out of Boundary Paths.py
"""
There is an m x n grid with a ball. The ball is initially at the position [startRow, startColumn]. You are allowed to move the ball to one of the four adjacent cells in the grid (possibly out of the grid crossing the grid boundary). You can apply at most maxMove moves to the ball.
Given the five integers m, n, maxMove, startRow, startColumn, return the number of paths to move the ball out of the grid boundary. Since the answer can be very large, return it modulo 109 + 7.

Example 1:

Input: m = 2, n = 2, maxMove = 2, startRow = 0, startColumn = 0
Output: 6

Example 2:

Input: m = 1, n = 3, maxMove = 3, startRow = 0, startColumn = 1
Output: 12
"""
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        @cache
        def helper(x, y, moves):
            if x == m or y == n or x < 0 or y < 0:
                return 1

            if moves == 0:
                return 0

            ans = helper(x - 1, y, moves - 1)
            ans += helper(x, y - 1, moves - 1)
            ans += helper(x + 1, y, moves - 1)
            ans += helper(x, y + 1, moves - 1)

            return ans % MOD

        MOD = 10**9 + 7

        return helper(startRow, startColumn, maxMove)

File: 576 Out of Boundary Paths/576 Out of Boundary Paths.go
var cords = []int{-1, 1}
var M = 1000000000 + 7

func findPathsRecursive(m, n, maxMove, startRow, startColumn int, cache [][][]int) int {

	// Check if row is out of bounds
	if startRow >= m || startRow < 0 {
		return 1
	}

	// Check if column is out of bounds
	if startColumn >= n || startColumn < 0 {
		return 1
	}

	// If no moves left, return
	if maxMove == 0 {
		return 0
	}

	if cache[startRow][startColumn][maxMove] != -1 {
		return cache[startRow][startColumn][maxMove]
	}

	paths := 0

	// Recurse for all adjacent coordinates
	for _, iCord := range cords {
		paths += findPathsRecursive(m, n, maxMove-1, startRow+iCord, startColumn, cache)
	}

	for _, jCord := range cords {
		paths += findPathsRecursive(m, n, maxMove-1, startRow, startColumn+jCord, cache)
	}

	cache[startRow][startColumn][maxMove] = paths % M

	return cache[startRow][startColumn][maxMove]

}

func findPaths(m int, n int, maxMove int, startRow int, startColumn int) int {

	// Initialise empty cache
	// cache[m][n][maxMove+1]
	cache := make([][][]int, m)
	for i := range cache {
		cache[i] = make([][]int, n)
		for j := range cache[i] {
			cache[i][j] = make([]int, maxMove+1)
			for k := range cache[i][j] {
				cache[i][j][k] = -1
			}
		}
	}

	pathCount := findPathsRecursive(m, n, maxMove, startRow, startColumn, cache)

	return pathCount

}
